gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriter.java

```diff
@@ -294,12 +298,14 @@ public void writeEnvelope(RecordEnvelope<GenericRecord> recordEnvelope) throws I
   }
 
   // Add fault tolerant ability and make sure we can emit GTE as desired
-  private void write(RecordEnvelope recordEnvelope, ConcurrentHashMap newSpecsMap, ConcurrentHashMap oldSpecsMap, HiveSpec spec) throws IOException {
+  @VisibleForTesting
+  void write(RecordEnvelope<GenericRecord> recordEnvelope, ConcurrentHashMap newSpecsMap, ConcurrentHashMap oldSpecsMap, HiveSpec spec) throws IOException {
     boolean meetException = false;
     String dbName = spec.getTable().getDbName();
     String tableName = spec.getTable().getTableName();
     String tableString = Joiner.on(TABLE_NAME_DELIMITER).join(dbName, tableName);
-    for (MetadataWriter writer : metadataWriters) {
+    List<MetadataWriter> allowedWriters = getAllowedMetadataWriters(toGMCE(recordEnvelope), metadataWriters);
```

Comment:
Can we avoid call toGMCE again in this method? As deepCopy is actually expansive..